### PR TITLE
refactor: merge entries and batch_index in raw_curp

### LINKED ISSUE #256: [Refactor]: Merge `entries` and `batch_index`
Currently in `RawCurp` all operations on `entries` and `batch_index` occur in pairs,`batch_index`'s length is always greater than `entries`'s length by 1 , to manage them needs to deal with the mismatch of these two `VecDeque`, I suggest merging them into the one `VecDeque` or specific data structure for unified management. maybe need to manage different length of them in specific data structure.
eg:
```Rust
self.entries.truncate(pi);
self.batch_index.truncate(pi.overflow_add(1));

self.batch_index.push_back(pre_entry_size.overflow_add(entry_size));
self.entries.push_back(entry);

let e = self.entries.pop_front().unwrap();
let _ig_i = self.batch_index.pop_front().unwrap();
```
after refactor:
```Rust
self.new_queue.truncate(pi);

self.new_queue.push_back(entry));

let e = self.new_queue.pop_front().unwrap();
```

### PR DESCRIPTION
**Pull request by liubog2008**
Close #256 

Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

* what changes does this pull request make?

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

### CODE REVIEW

**Comment by Phoenix500526** [comment]
@liubog2008 Thank you for your contribution. Some CI jobs in this pr have failed. Please perform a self-test before submitting.  For specific requirements, please refer to the [CONTRIBUTING.md doc](https://github.com/xline-kv/Xline/blob/master/CONTRIBUTING.md#pull-requests). A better way to do a self-test is to use the pre-commit tool. Xline provides a [pre-commit configuration file](https://github.com/xline-kv/Xline/blob/master/.pre-commit-config.yaml) to install some pre-commit hooks and commit-msg hooks. You can install these hooks by executing `pre-commit install --hook-type pre-commit --hook-type commit-msg·.

**Comment by Phoenix500526** [comment]
Please squash these two commit as one and add a Signed-off-by line in the commit message.

**Comment by liubog2008** [comment]
I can't reproduce the failed case in my local side and I think the failure is not introduced by this PR?

File: curp/src/server/raw_curp/log.rs [hunk 1]
```diff
@@ -43,14 +43,135 @@ pub(super) struct Log<C: Command> {
...
+    /// push a log entry into the back of queue
+    fn push_back(&mut self, entry: LogEntry<C>) -> Result<(), bincode::Error> {
+        let entry_size = serialized_size(&entry)?;
+
+        self.entries.push_back(entry);
+        let pre_entry_size = if let Some(&last_entry_size) = self.batch_index.back() {
+            last_entry_size
+        } else {
+            0
+        };
```
Comment (themanforfree) [inline curp/src/server/raw_curp/log.rs:94]: Return an error when `batch_index.back()` is `None`? it should always be `Some`

File: curp/src/server/raw_curp/log.rs [hunk 2]
```diff
@@ -43,14 +43,135 @@ pub(super) struct Log<C: Command> {
     pub(super) last_exe: LogIndex,
     /// Tx to send log entries to persist task
     log_tx: mpsc::UnboundedSender<LogEntry<C>>,
-    /// Prefix sum vector of log entries
-    batch_index: VecDeque<u64>,
     /// Batch size limit
     batch_limit: u64,
```
Comment (themanforfree) [inline curp/src/server/raw_curp/log.rs:47]: `batch_limit` just used in the `LogEntryVecDeque`, I recommend move it to `LogEntryVecDeque`.

**Comment by bsbds** [comment]
> I can't reproduce the failed case in my local side and I think the failure is not introduced by this PR?

Sorry about the confustion, it seems there's a bug in the test. I'll resolve this in another PR.

File: curp/src/server/raw_curp/log.rs [hunk 3]
```diff
@@ -43,14 +43,135 @@ pub(super) struct Log<C: Command> {
...
+        batch_index.push_back(0);
+        Self {
+            /// Now keep the old logic to init entries with LOG_INIT_CAPACITY
+            /// TODO: maybe use the cap in args to init vec?
+            /// - entries: cap ?
+            /// - batch_index: cap + 1 ?
+            entries: VecDeque::with_capacity(LOG_INIT_CAPACITY),
+            batch_index,
+        }
+    }
```
[the remark this answers is not in the document]
Comment (liubog2008) [reply curp/src/server/raw_curp/log.rs:75]: There is an arg called `entries_cap` in `Log::new`. It seems that `LOG_INIT_CAPACITY` can be removed?

File: curp/src/server/raw_curp/log.rs (hunk 1 above)
Comment (liubog2008) [reply curp/src/server/raw_curp/log.rs:94]: I think it's not reasonable to throw an error, how about `panic!` at `else` branch? It must be a programmatic bug if `back()` is `None`.

File: curp/src/server/raw_curp/log.rs (hunk 2 above)
Comment (liubog2008) [reply curp/src/server/raw_curp/log.rs:47]: lgtm

File: curp/src/server/raw_curp/log.rs (hunk 1 above)
Comment (Phoenix500526) [reply curp/src/server/raw_curp/log.rs]: How about `let pre_entries_size = self.batch_index.back().unwrap_or_else(|| unreachable!("batch_index.back() should never be None"));`

Comment (Phoenix500526) [reply curp/src/server/raw_curp/log.rs]: How about `let pre_entries_size = self.batch_index.back().unwrap_or_else(|| unreachable!("batch_index.back() should never be None"));`

**Review by themanforfree — APPROVED** [review]
LGTM

**Comment by iGxnon** [comment]
Please rebase based on the latest commit, as some CI steps have changed. This may be the reason why the `Test Normal` job was inexplicably canceled.

**Review by Phoenix500526 — APPROVED** [review]
LGTM

**Review by iGxnon — APPROVED** [review]
LGTM

**Pull request merged**